firestore.rules

```diff
@@ -15,62 +15,175 @@ service cloud.firestore {
       allow read: if true;
     }
 
+    //
+    //  -----------  Student Work -------------
+    //
+
+    function teacherOfContext() {
+      return request.auth.token.user_type == 'teacher'
+        && request.auth.token.class_hash == resource.data.context_id
+        && request.auth.token.platform_id == resource.data.platform_id;
+    }
+
+    // Note: this conditional is used in read rules, which do not require a
+    // matching context_id, so the context_id is not enforced here
+    function learnerOwner(res) {
+      return request.auth.token.user_type == 'learner'
+        && string(request.auth.token.platform_user_id) == string(res.data.platform_user_id)
+        && request.auth.token.platform_id == res.data.platform_id;
+    }
+
+    function notChangingIdentifyingProps() {
+      // Make sure the learner isn't changing the document's identifying properties
+      return string(request.resource.data.platform_user_id) == string(request.auth.token.platform_user_id)
+        && request.resource.data.platform_id == request.auth.token.platform_id
+        // no changing the document's context outside of the learner's context
+        && request.resource.data.context_id == request.auth.token.class_hash
+        // no changing the document's context from what it was before
+        // this could only happen if learner got access to an existing document in a different context
+        && request.resource.data.context_id == resource.data.context_id;
+        // This implicitly also means that
+        // request.auth.token.class_hash == resource.data.context_id
+    }
+
+    // Check the owner of the request.resource and
+    // enforce the context_id matches the auth info
+    function learnerOwnerCreate() {
+      return learnerOwner(request.resource)
+        // the platform_id is checked in the learnerOwner check
+        && request.auth.token.class_hash == request.resource.data.context_id;
+    }
+
+    function anonymousCreate() {
+      return request.resource.data.run_key.size() > 10
+        // Make sure anonymous answers don't have platform_ids
+        && !('platform_id' in request.resource.data);
+    }
+
+    // Anyone can read any anonymous docs with a valid run_key
+    function anonymousRead() {
+      return resource.data.run_key.size() > 10
+    }
+
+    function anonymousUpdate() {
+      return resource.data.run_key.size() > 10
+        // Make sure they can't change the run key of an existing document
+        && resource.data.run_key == request.resource.data.run_key
+        // Make sure anonymous answers don't have platform_ids
+        && !('platform_id' in request.resource.data);
+    }
+
+    function studentWorkCreate() {
+      return anonymousCreate()
+        || learnerOwnerCreate();
+    }
+
+    function studentWorkRead() {
+      return anonymousRead()
+        || teacherOfContext()
+        || learnerOwner(resource);
+    }
+
+    function studentWorkUpdate() {
+      return anonymousUpdate()
```

Comment:
I would think that `notChangingIdentifyingProps()` would apply to the `anonymousUpdate()` as well.

Reply:
The concept applies, but that actual function is specific to the authenticated identifying props of platform_id and context_id. For anonymous the only identifying prop is the run_key, and also make sure they don't set a platform. So the anonymousUpdate checks those things itself.